### PR TITLE
[TF FE] Support complex tensors for ReverseSequence operations

### PR DESCRIPTION
**Pull request by MonalSD**
### Details:
 - *Extend the loader for  ReverseSequence operations on complex type tensors*
 - Addressed #23236

### CODE REVIEW

File: src/frontends/tensorflow_common/src/op/reverse_sequence.cpp [hunk 1]
```diff
@@ -14,14 +16,29 @@ namespace frontend {
...
 
     // retrieve attributes
     auto seq_dim = node.get_attribute<int64_t>("seq_dim");
     auto batch_dim = node.get_attribute<int64_t>("batch_dim", 0);
 
+    auto complex_type_mark = as_type_ptr<ComplexTypeMark>(input.get_node_shared_ptr());
+    if (complex_type_mark) {
+        auto const_one = make_shared<v0::Constant>(element::i32, Shape{1}, 1);
+        auto updated_seq_dim = make_shared<v1::Add>(seq_dim, const_one);
+        auto updated_batch_dim = make_shared<v1::Add>(batch_dim, const_one);
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/reverse_sequence.cpp:31]: not clear why are `seq_dim` and `batch_dim` incremented. For non-negative values, there is no update. For negative values, we should add `rank-1`, where `rank` is a rank of `input`

Please add tests with negative axes.

**Comment by MonalSD** [comment]
@rkazants addressed the changes. Please give it a look

File: src/frontends/tensorflow_common/src/op/reverse_sequence.cpp (hunk 1 above)
Comment (rkazants) [reply src/frontends/tensorflow_common/src/op/reverse_sequence.cpp:38]: Not correct because `seq_dim` is of `int64_t` type and you pass it into Graph operation `Add`. You should update it using `Select` and `Rank` operations

File: src/frontends/tensorflow_common/src/op/reverse_sequence.cpp [hunk 2]
```diff
@@ -14,13 +16,35 @@ namespace frontend {
...
     // retrieve attributes
     auto seq_dim = node.get_attribute<int64_t>("seq_dim");
     auto batch_dim = node.get_attribute<int64_t>("batch_dim", 0);
+    // Handling negative values by adding rank if needed
+    if (seq_dim < 0) {
+        seq_dim += input.get_shape().size();
+    }
+    if (batch_dim < 0) {
+        batch_dim += input.get_shape().size();
+    }
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/reverse_sequence.cpp:32]: Not correct because seq_dim is of int64_t type and you pass it into Graph operation Add. You should update it using Select and Rank operations

Comment (MonalSD) [reply src/frontends/tensorflow_common/src/op/reverse_sequence.cpp:32]: @rkazants cool I'll make the changes. Thanks for input

**Comment by MonalSD** [comment]
@rkazants addressed the changes please give your feedback.

File: src/frontends/tensorflow_common/src/op/reverse_sequence.cpp [hunk 3]
```diff
@@ -14,13 +19,47 @@ namespace frontend {
...
+
+    auto complex_type_mark = as_type_ptr<ComplexTypeMark>(input.get_node_shared_ptr());
+    if (complex_type_mark) {
+        auto const_one = make_shared<v0::Constant>(element::i32, Shape{1}, 1);
+        auto seq_dim_tensor = make_shared<v1::Select>(
+            make_shared<v1::Equal>(make_shared<v1::Rank>(input), const_one),
+            make_shared<v0::Constant>(element::i64, Shape{}, seq_dim),
+            make_shared<v1::Select>(make_shared<v1::Equal>(make_shared<v1::Rank>(input), const_one),
+                                    make_shared<v1::Subtract>(make_shared<v1::Rank>(input), const_one),
+                                    make_shared<v0::Constant>(element::i64, Shape{}, -1)));
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/reverse_sequence.cpp:45]: not clear logic.
please create helper function to update `seq_dim` and `batch_dim` because the logic is the same.
please create separate variable for each step like `auto rank =...` and `auto minus_one_const=...` etc. This way we can clarify what the logic is meant here.

File: tests/layer_tests/tensorflow_tests/test_tf_ReverseSequence.py [hunk 4]
```diff
@@ -51,3 +51,58 @@ def test_reverse_sequence_basic(self, params, ie_device, precision, ir_version,
...
+        return tf_net, None
+
+    test_data_basic = [
+        dict(input_shape=[2, 3], input_type=np.int32, seq_lengths_type=np.int64, seq_dim=1, batch_dim=0),
+        dict(input_shape=[3, 6, 4], input_type=np.float32, seq_lengths_type=np.int32, seq_dim=2, batch_dim=0),
+        dict(input_shape=[6, 3, 4, 2], input_type=np.float32, seq_lengths_type=np.int32, seq_dim=0, batch_dim=3),
+    ]
+
+    @pytest.mark.parametrize("params", test_data_basic)
+    @pytest.mark.precommit_tf_fe
```
Comment (rkazants) [inline tests/layer_tests/tensorflow_tests/test_tf_ReverseSequence.py:101]:
```suggestion
    @pytest.mark.precommit
```

File: src/frontends/tensorflow_common/src/op/reverse_sequence.cpp [hunk 5]
```diff
@@ -14,13 +19,47 @@ namespace frontend {
...
+        auto batch_dim_tensor = make_shared<v1::Select>(
+            make_shared<v1::Equal>(make_shared<v1::Rank>(input), const_one),
+            make_shared<v0::Constant>(element::i64, Shape{}, batch_dim),
+            make_shared<v1::Select>(make_shared<v1::Equal>(make_shared<v0::Rank>(input), const_one),
+                                    make_shared<v1::Subtract>(make_shared<v1::Rank>(input), const_one),
+                                    make_shared<v0::Constant>(element::i64, Shape{}, -1)));
+        auto updated_seq_dim = make_shared<v1::Add>(seq_dim_tensor, const_one);
+        auto updated_batch_dim = make_shared<v1::Add>(batch_dim_tensor, const_one);
+
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/reverse_sequence.cpp:54]:
```suggestion
```

File: src/frontends/tensorflow_common/src/op/reverse_sequence.cpp [hunk 6]
```diff
@@ -14,13 +19,31 @@ namespace frontend {
...
     // retrieve attributes
     auto seq_dim = node.get_attribute<int64_t>("seq_dim");
     auto batch_dim = node.get_attribute<int64_t>("batch_dim", 0);
+    // Handling negative values by adding rank if needed
+    if (seq_dim < 0) {
+        seq_dim += input.get_shape().size();
+    }
+    if (batch_dim < 0) {
+        batch_dim += input.get_shape().size();
+    }
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/reverse_sequence.cpp:35]:
```suggestion
    // handling negative values
    if (seq_dim < 0) {
        seq_dim -= 1;
    }
    if (batch_dim < 0) {
        batch_dim -= 1;
    }
```

File: tests/layer_tests/tensorflow_tests/test_tf_ReverseSequence.py [hunk 7]
```diff
@@ -51,3 +51,58 @@ def test_reverse_sequence_basic(self, params, ie_device, precision, ir_version,
...
+            tf.raw_ops.ReverseSequence(input=input, seq_lengths=seq_lengths, seq_dim=seq_dim, batch_dim=batch_dim)
+            tf.compat.v1.global_variables_initializer()
+            tf_net = sess.graph_def
+
+        return tf_net, None
+
+    test_data_basic = [
+        dict(input_shape=[2, 3], input_type=np.int32, seq_lengths_type=np.int64, seq_dim=1, batch_dim=0),
+        dict(input_shape=[3, 6, 4], input_type=np.float32, seq_lengths_type=np.int32, seq_dim=2, batch_dim=0),
+        dict(input_shape=[6, 3, 4, 2], input_type=np.float32, seq_lengths_type=np.int32, seq_dim=0, batch_dim=3),
```
Comment (rkazants) [inline tests/layer_tests/tensorflow_tests/test_tf_ReverseSequence.py:97]: please add negative cases for `seq_dim` and `batch_dim`

**Comment by rkazants** [comment]
build_jenkins

File: src/frontends/tensorflow_common/src/op/reverse_sequence.cpp [hunk 8]
```diff
@@ -14,14 +19,45 @@ namespace frontend {
...
+            make_shared<v8::Gather>(base_input, gather_index_real, v0::Constant::create(element::i32, Shape{}, {-1}));
+        auto input_imag =
+            make_shared<v8::Gather>(base_input, gather_index_imag, v0::Constant::create(element::i32, Shape{}, {-1}));
+
+        // Reverse sequence for real and imaginary parts
+        auto reversed_real = make_shared<v0::ReverseSequence>(input_real, seq_lengths, batch_dim, seq_dim);
+        auto reversed_imag = make_shared<v0::ReverseSequence>(input_imag, seq_lengths, batch_dim, seq_dim);
+
+        // Recombine into a complex tensor
+        auto complex_concat = make_shared<v0::Concat>(OutputVector{reversed_real, reversed_imag}, -1);
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/reverse_sequence.cpp:54]: not needed to split into real and imag parts. Please pass `base_input` to `ReverseSequence`

Comment (MonalSD) [reply src/frontends/tensorflow_common/src/op/reverse_sequence.cpp:54]: Okay will resolve it

**Comment by MonalSD** [comment]
@rkazants hey review changes.

File: src/frontends/tensorflow_common/src/op/reverse_sequence.cpp [hunk 9]
```diff
@@ -14,14 +19,36 @@ namespace frontend {
...
+    auto complex_type_mark = as_type_ptr<ComplexTypeMark>(input.get_node_shared_ptr());
+    if (complex_type_mark) {
+        auto base_input = complex_type_mark->input_value(0);
+        // Reverse sequence for real and imaginary parts
+        auto reverse_sequence = make_shared<v0::ReverseSequence>(base_input, seq_lengths, batch_dim, seq_dim);
+
+        auto complex_result =
+            make_shared<ComplexTypeMark>(reverse_sequence, complex_type_mark->get_complex_part_type());
+
+        set_node_name(node.get_name(), complex_result);
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/reverse_sequence.cpp:47]: please `set_node_name` to `reverse_sequence`. It is not worth to set for `complex_result` that will be removed eventually.

**Comment by MonalSD** [comment]
@rkazants please review

File: src/frontends/tensorflow_common/src/op/reverse_sequence.cpp [hunk 10]
```diff
@@ -14,14 +19,36 @@ namespace frontend {
...
+
+    auto complex_type_mark = as_type_ptr<ComplexTypeMark>(input.get_node_shared_ptr());
+    if (complex_type_mark) {
+        auto base_input = complex_type_mark->input_value(0);
+        // Reverse sequence for real and imaginary parts
+        auto reverse_sequence = make_shared<v0::ReverseSequence>(base_input, seq_lengths, batch_dim, seq_dim);
+
+        auto complex_result =
+            make_shared<ComplexTypeMark>(reverse_sequence, complex_type_mark->get_complex_part_type());
+
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/reverse_sequence.cpp:46]:
```suggestion
        auto complex_result =
            make_shared<ComplexTypeMark>(reverse_sequence, complex_type_mark->get_complex_part_type());
```

**Comment by rkazants** [comment]
build_jenkins

**Comment by rkazants** [comment]
build_jenkins
